File: objectiveai-sdk-py/objectiveai_sdk/stream.py
"""Server-Sent Events (SSE) stream for ObjectiveAI API responses."""

from __future__ import annotations

import json
from typing import Any, AsyncIterator, Generic, TypeVar

from objectiveai_sdk.error import ObjectiveAIFetchError, _is_response_error

T = TypeVar("T")


class Stream(Generic[T]):
    """Async iterator over Server-Sent Events from an API response.

    Parses SSE-formatted data from an ``httpx`` streaming response and
    yields parsed JSON chunks.  Raises :class:`ObjectiveAIFetchError` if
    an error event is received in the stream.

    Usage::

        async with client.post_streaming("/path", body) as stream:
            async for chunk in stream:
                print(chunk)
    """

    def __init__(self, response: Any) -> None:  # httpx.Response
        self._response = response
        self._buffer = ""
        self._done = False

    async def __aiter__(self) -> AsyncIterator[T]:
        try:
            async for raw_bytes in self._response.aiter_bytes():
                if self._done:
                    break

                self._buffer += raw_bytes.decode("utf-8", errors="replace")

                # Split by double newline (SSE event separator)
                parts = self._buffer.split("\n\n")

                # Keep the last part in the buffer (might be incomplete)
                self._buffer = parts.pop()

                for part in parts:
                    for event in self._parse_sse(part):
                        yield event
                        if self._done:
                            return

            # Process any remaining data in buffer
            if self._buffer.strip():
                for event in self._parse_sse(self._buffer):
                    yield event
        finally:
            await self._response.aclose()

    def _parse_sse(self, text: str) -> list[T]:
        """Parse an SSE event block and return parsed data events."""
        results: list[T] = []
        for line in text.split("\n"):
            if not line or line.startswith(":"):
                continue

            if line.startswith("data:"):
                data = line[5:].strip()

                if data == "[DONE]":
                    self._done = True
                    continue

                if not data:
                    continue

                parsed = json.loads(data)

                if _is_response_error(parsed):
                    raise ObjectiveAIFetchError(parsed)

                results.append(parsed)

        return results

    async def to_list(self) -> list[T]:
        """Collect all events into a list."""
        items: list[T] = []
        async for item in self:
            items.append(item)
        return items

File: objectiveai-sdk-js/src/stream.ts
import { ObjectiveAIFetchError, isResponseError } from "./error";

/**
 * A readable stream wrapper for Server-Sent Events (SSE).
 * Provides async iteration over parsed SSE data events.
 * Throws ObjectiveAIFetchError if the API returns an error in the stream.
 *
 * Two backing sources are supported:
 *
 * 1. `new Stream(response, controller?)` — wraps an SSE-emitting
 *    HTTP response; this is the path used by `fetch`-mode clients.
 *
 * 2. `Stream.fromAsyncIterable(iter, controller?)` — wraps a
 *    pre-parsed AsyncIterable of chunks; used by `viewer: true`
 *    clients so the consumer-facing `Stream<T>` type stays identical
 *    across both fetch and Tauri-postMessage backends.
 */
export class Stream<T> implements AsyncIterable<T> {
  private reader: ReadableStreamDefaultReader<Uint8Array> | null;
  private decoder: TextDecoder;
  private buffer: string = "";
  private done: boolean = false;
  private controller: AbortController | null;
  private asyncIterableSource: AsyncIterable<T> | null;

  constructor(response: Response, controller?: AbortController | null) {
    if (!response.body) {
      throw new Error("Response body is null");
    }
    this.reader = response.body.getReader();
    this.decoder = new TextDecoder();
    this.controller = controller ?? null;
    this.asyncIterableSource = null;
  }

  /**
   * Wrap an `AsyncIterable<T>` (e.g. the viewer-mode api-call
   * iterator) as a `Stream<T>`. Skips the SSE parsing path entirely;
   * the iterator yields chunks pre-parsed by its source.
   */
  static fromAsyncIterable<T>(
    source: AsyncIterable<T>,
    controller?: AbortController | null,
  ): Stream<T> {
    const stream = Object.create(Stream.prototype) as Stream<T>;
    stream.reader = null;
    stream.decoder = new TextDecoder();
    stream.buffer = "";
    stream.done = false;
    stream.controller = controller ?? null;
    stream.asyncIterableSource = source;
    return stream;
  }

  /**
   * Abort the stream.
   */
  abort(): void {
    this.controller?.abort();
  }

  async *[Symbol.asyncIterator](): AsyncIterator<T> {
    if (this.asyncIterableSource) {
      for await (const chunk of this.asyncIterableSource) {
        yield chunk;
      }
      return;
    }

    if (!this.reader) {
      throw new Error("Stream has no backing source");
    }
    const reader = this.reader;
    try {
      while (!this.done) {
        const { value, done } = await reader.read();

        if (done) {
          this.done = true;
          // Process any remaining data in buffer
          if (this.buffer.trim()) {
            const events = this.parseSSE(this.buffer);
            for (const event of events) {
              if (event !== null) {
                yield event;
              }
            }
          }
          break;
        }

        this.buffer += this.decoder.decode(value, { stream: true });

        // Split by double newline (SSE event separator)
        const parts = this.buffer.split(/\n\n/);

        // Keep the last part in the buffer (might be incomplete)
        this.buffer = parts.pop() ?? "";

        // Process complete events
        for (const part of parts) {
          const events = this.parseSSE(part);
          for (const event of events) {
            if (event !== null) {
              yield event;
            }
          }
        }
      }
    } finally {
      reader.releaseLock();
    }
  }

  /**
   * Parse SSE format and extract data.
   * Returns null for [DONE] or empty events.
   *
   * SSE format:
   * - Lines starting with `:` are comments and are ignored
   * - `data:` lines contain the event payload
   * - Empty lines separate events
   * - We only process `data:` fields, ignoring `event:`, `id:`, `retry:`
   */
  private parseSSE(text: string): (T | null)[] {
    const results: (T | null)[] = [];
    const lines = text.split("\n");
    let hasDataLine = false;

    for (const line of lines) {
      // Skip empty lines
      if (!line) {
        continue;
      }

      // Skip comment lines (lines starting with `:`)
      if (line.startsWith(":")) {
        continue;
      }

      // Parse "data: ..." lines
      if (line.startsWith("data:")) {
        hasDataLine = true;
        const data = line.slice(5).trim();

        // Check for stream termination
        if (data === "[DONE]") {
          this.done = true;
          continue;
        }

        // Skip empty data
        if (!data) {
          continue;
        }

        const parsed = JSON.parse(data);

        // Check if this is a ResponseError
        if (isResponseError(parsed)) {
          throw new ObjectiveAIFetchError(parsed);
        }

        results.push(parsed as T);
      }
      // Ignore other SSE fields like event:, id:, retry:
    }

    // If the event block contained only comments (no data lines), return empty
    if (!hasDataLine && results.length === 0) {
      return [];
    }

    return results;
  }

  /**
   * Collect all events into an array.
   */
  async toArray(): Promise<T[]> {
    const results: T[] = [];
    for await (const item of this) {
      results.push(item);
    }
    return results;
  }
}
